ipalib/x509.py

```diff
@@ -543,3 +545,28 @@ def format_datetime(t):
     if t.tzinfo is None:
         t = t.replace(tzinfo=UTC())
     return unicode(t.strftime("%a %b %d %H:%M:%S %Y %Z"))
+
+
+def match_hostname(cert, hostname):
+    match_cert = {}
+
+    match_cert['subject'] = match_subject = []
+    for rdn in cert.subject.rdns:
+        match_rdn = []
+        for ava in rdn:
+            if ava.oid == cryptography.x509.oid.NameOID.COMMON_NAME:
+                match_rdn.append(('commonName', ava.value))
+        match_subject.append(match_rdn)
+
+    try:
+        san = cert.extensions.get_extension_for_class(
+                cryptography.x509.SubjectAlternativeName)
+    except cryptography.x509.ExtensionNotFound:
+        pass
+    else:
+        match_cert['subjectAltName'] = match_san = []
+        for value in san.value.get_values_for_type(cryptography.x509.DNSName):
+            value = DNSName(value).ToASCII()
```

Comment:
You must not our DNSName class here. It's an invalid and therefore unsafe conversion. I fear we cannot convert back safely. In general IDNA is not a bijective encoding. This is especially true for different kinds of IDNA. PyCA cryptography uses IDNA 2008 UTS46, DNSName uses IDNA 2003.

Consider this example:

```
>>> from ipapython.dnsutil import DNSName
>>> import idna
>>> asn1_string = 'xn--strae-oqa.de'
>>> pyca_dnsname = idna.decode(asn1_string)
>>> DNSName(pyca_dnsname).ToASCII()
u'strasse.de'
```

I talked to @reaperhulk about the issue. Paul pointed me to https://github.com/pyca/cryptography/issues/3357.

Yes, hostname verification is a mess. I got it wrong multiple times, caused at least one CVE and found multiple other CVEs. :/